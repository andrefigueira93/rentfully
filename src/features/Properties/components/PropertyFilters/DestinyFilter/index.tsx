import { Button } from "@/components/ui/button";
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
} from "@/components/ui/command";
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover";
import { cn } from "@/lib/utils";
import { locationFilterAtom } from "@/store/booking-filters-store";
import { uniqueLocationsAtom } from "@/store/properties-store";
import { CheckIcon, ChevronDownIcon } from "@radix-ui/react-icons";
import { useAtom, useAtomValue } from "jotai";
import React, { useMemo } from "react";
import ClearButton from "../ClearButton";
import FilterWrapper from "../FilterWrapper";

export default function DestinyFilter() {
  const locations = useAtomValue(uniqueLocationsAtom);
  const [value, setValue] = useAtom(locationFilterAtom);
  const [open, setOpen] = React.useState(false);
  const mapedLocations = useMemo(
    () =>
      locations.map((l) => ({
        label: l?.city ?? "",
        value: l?.city ?? "",
      })),
    [locations]
  );

  function handleClear() {
    setValue(undefined);
  }

  return (
    <FilterWrapper label="Destiny">
      <Popover open={open} onOpenChange={setOpen}>
        <PopoverTrigger asChild>
          <Button
            variant="outline"
            role="combobox"
            aria-expanded={open}
            className="justify-between w-full"
          >
            {value
              ? mapedLocations.find(
                  (framework) =>
                    framework.value?.toLowerCase() === value?.toLowerCase()
                )?.label
              : "Pick a location..."}
            <ChevronDownIcon className="ml-2 h-4 w-4 shrink-0 opacity-50" />
          </Button>
        </PopoverTrigger>
        <PopoverContent className="w-[200px] max-h-60 overflow-scroll p-0">
          <Command>
            <CommandInput placeholder="Pick a location..." />
            <CommandEmpty>Destiny not found.</CommandEmpty>
            <CommandGroup>
              {mapedLocations.map((framework, i) => (
                <CommandItem
                  data-testid={`location-option-${i}`}
                  key={framework.value}
                  value={framework.value}
                  onSelect={(currentValue) => {
                    setValue(currentValue === value ? "" : currentValue);
                    setOpen(false);
                  }}
                >
                  <CheckIcon
                    className={cn(
                      "mr-2 h-4 w-4",
                      value === framework.value ? "opacity-100" : "opacity-0"
                    )}
                  />
                  {framework.label}
                </CommandItem>
              ))}
            </CommandGroup>
          </Command>
        </PopoverContent>
      </Popover>
      {value && <ClearButton onClick={handleClear} />}
    </FilterWrapper>
  );
}
