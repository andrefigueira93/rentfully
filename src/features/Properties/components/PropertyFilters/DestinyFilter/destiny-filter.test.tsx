/* eslint-disable @typescript-eslint/no-explicit-any */
import Providers from "@/providers";
import { fireEvent, render, screen } from "@testing-library/react";
import { useAtom } from "jotai";
import { MemoryRouter } from "react-router";
import DestinyFilter from "./index";

vi.mock("jotai", async () => {
  const originalModule = await import("jotai");

  return {
    ...originalModule,
    useAtom: vi.fn(),
  };
});
const mockSetValue = vi.fn() as never;
const selectedLocation = "new york";

describe("DestinyFilter Component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useAtom).mockReturnValue([selectedLocation, mockSetValue]);
  });

  test("renders correctly", () => {
    const { container } = render(
      <MemoryRouter>
        <Providers>
          <DestinyFilter />
        </Providers>
      </MemoryRouter>
    );

    expect(screen.getByText("Destiny")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  // test('opens and closes the popover', () => {
  //   const { container } = render(
  //     <MemoryRouter>
  //       <Providers>
  //         <DestinyFilter />
  //       </Providers>
  //     </MemoryRouter>
  //   );

  //   const button = screen.getByRole('combobox');
  //   fireEvent.click(button);

  //   expect(screen.getByRole('listbox')).toBeInTheDocument();

  //   fireEvent.click(button);

  //   expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  //   expect(container).toMatchSnapshot();
  // });

  test("selects a location", () => {
    render(
      <MemoryRouter>
        <Providers>
          <DestinyFilter />
        </Providers>
      </MemoryRouter>
    );

    const button = screen.getByRole("combobox");
    fireEvent.click(button);

    const locationOption = screen.getByTestId("location-option-0");
    fireEvent.click(locationOption);

    expect(mockSetValue).toHaveBeenCalled();

    // expect(container).toMatchSnapshot();
    // Can't make snapshot because of the random generation
  });

  test("displays selected destiny", () => {
    render(
      <MemoryRouter>
        <Providers>
          <DestinyFilter />
        </Providers>
      </MemoryRouter>
    );

    const button = screen.getByRole("combobox");
    fireEvent.click(button);

    const locationOption = screen.getByTestId("location-option-0");
    fireEvent.click(locationOption);

    const text = locationOption.textContent;

    expect(screen.findAllByText(text as string)).toBeTruthy();

    // Can't make snapshot because of the random generation
    // of the options, but it works.
  });

  test("clears the selected location", () => {
    const selectedLocation = "New York";
    const setValueMock = vi.fn() as never;
    vi.mocked(useAtom).mockReturnValue([selectedLocation, setValueMock]);

    // Can't make snapshot because of the random generation
    // of the options, but it works.

    // If you wanna check the snapshot, uncomment the line below
    // const { container } =
    render(
      <MemoryRouter>
        <Providers>
          <DestinyFilter />
        </Providers>
      </MemoryRouter>
    );

    const clearButton = screen.getByTestId("clear-button");
    fireEvent.click(clearButton);

    expect(setValueMock).toHaveBeenCalledWith(undefined);
    // If you wanna check the snapshot, uncomment the line below
    // expect(container).toMatchSnapshot();
  });
});
