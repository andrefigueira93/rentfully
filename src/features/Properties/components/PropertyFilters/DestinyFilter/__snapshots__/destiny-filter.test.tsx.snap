// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DestinyFilter Component > renders correctly 1`] = `
<div>
  <div
    class="flex flex-col space-y-2"
  >
    <div
      class="w-full flex items-center justify-evenly"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-full"
      >
        Destiny
      </label>
    </div>
    <div
      class="w-full flex items-center justify-evenly"
    >
      <button
        aria-controls="radix-:r0:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2 justify-between w-full"
        data-state="closed"
        role="combobox"
        type="button"
      >
        <svg
          class="ml-2 h-4 w-4 shrink-0 opacity-50"
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3.13523 6.15803C3.3241 5.95657 3.64052 5.94637 3.84197 6.13523L7.5 9.56464L11.158 6.13523C11.3595 5.94637 11.6759 5.95657 11.8648 6.15803C12.0536 6.35949 12.0434 6.67591 11.842 6.86477L7.84197 10.6148C7.64964 10.7951 7.35036 10.7951 7.15803 10.6148L3.15803 6.86477C2.95657 6.67591 2.94637 6.35949 3.13523 6.15803Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground dark:text-white h-9 px-4 py-2 ml-1"
        data-testid="clear-button"
      >
        <svg
          fill="none"
          height="15"
          viewBox="0 0 15 15"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.8536 2.85355C13.0488 2.65829 13.0488 2.34171 12.8536 2.14645C12.6583 1.95118 12.3417 1.95118 12.1464 2.14645L7.5 6.79289L2.85355 2.14645C2.65829 1.95118 2.34171 1.95118 2.14645 2.14645C1.95118 2.34171 1.95118 2.65829 2.14645 2.85355L6.79289 7.5L2.14645 12.1464C1.95118 12.3417 1.95118 12.6583 2.14645 12.8536C2.34171 13.0488 2.65829 13.0488 2.85355 12.8536L7.5 8.20711L12.1464 12.8536C12.3417 13.0488 12.6583 13.0488 12.8536 12.8536C13.0488 12.6583 13.0488 12.3417 12.8536 12.1464L8.20711 7.5L12.8536 2.85355Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
